### PR TITLE
CT-2960 CT-3005 edit and data requests table

### PR DESCRIPTION
**Pull request by Polsonby**
## Description
This PR includes the work to display and edit data requests, as well as mark them complete. 

It also includes changes to the data table design that emerged as late requirements following user testing and feedback. 

You can:
* View the data request details entered on the New Data Request form using the new request_types and from/to dates when relevant
* Edit a data request to amend the location, request type, date requested, and from/to dates
* Edit a data request to update the number of pages received
* Edit a data request to update the status to completed 

Some things to observe
1. The status outputs as "Completed" or "In progress" depending on the completed status value
2. The from/to dates output in a row beneath the request type if they're present
3. If the from date is omitted, the line reads "Up to" the _to date_
4. If the to date is omitted, the line reads _from date_ "onwards"
5. When you edit the details of the data request, the Case History shows a generic status line
6. When you edit the pages received, the Case history shows a custom message with the old and new page counts
7. When you choose "Edit" on a data request and then save without any changes, the case show page reloads with a notice to say no changes were made and NO message appears in the case history

The data request log history is no longer updated when the page count changes. This is a known issue and there is no intent to fix it right now. When we come to the Chasing stories we will revisit the data request logs functionality. 

## Self-review checklist

* [x] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [x] (3) Tests passing
* [x] (4) Branch ready to be merged (not work in progress)
* [ ] (5) No superfluous changes in diff
* [x] (6) No TODO's without new ticket numbers
* [x] (7) PR Prefixed with ticket number e.g. `CT-7654 ...`

### Screenshots

image

image

image

### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-3005
https://dsdmoj.atlassian.net/browse/CT-2960
https://dsdmoj.atlassian.net/browse/CT-2993

### Deployment
This PR includes database migrations that can be rolled back if necessary

`rails db:migrate`

### Manual testing instructions
Run the database migrations
Edit an existing Offender SAR
Create a new data request
Verify the details output appropriately
Verify the edit function allows you to edit the details
Verify the complete checkbox updates the status on the table
Verify the pages received mechanism works as intended

### CODE REVIEW

File: db/migrate/20200819171428_drop_and_recreate_request_type_enum.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class DropAndRecreateRequestTypeEnum < ActiveRecord::Migration[5.2]
+  def up
+    execute <<-SQL
+      UPDATE data_requests SET request_type = 'all_prison_records';
```
Comment (ymao2) [inline db/migrate/20200819171428_drop_and_recreate_request_type_enum.rb:4]: This line destroyed my local data :)

File: app/controllers/cases/data_requests_controller.rb [hunk 2]
```diff
@@ -53,9 +50,9 @@ def update
         redirect_to case_path(@case)
       when :unprocessed
         flash[:notice] = t('.unprocessed')
-        redirect_to edit_case_data_request_path(@case, @data_request)
+        redirect_to case_path(@case)
       when :error
-        @data_request_log = service.data_request_log
+        @data_request = service.data_request
```
Comment (ymao2) [inline app/controllers/cases/data_requests_controller.rb:55]: I just realised you removed all those parts related to data_request_logs,   in that case I don't think I need to insert any data into this table

File: spec/controllers/cases/data_requests_controller_spec.rb [hunk 3]
```diff
@@ -77,12 +83,11 @@
       }
     }
 
-    it 'builds a new data_request_log with last received values' do
+    it 'builds a new data_request with last received values' do
       get :edit, params: params
 
-      expect(assigns(:data_request_log).new_record?).to eq true
-      expect(assigns(:data_request_log).num_pages).to eq 10
-      expect(assigns(:data_request_log).date_received).to eq Date.yesterday
```
Comment (ymao2) [inline spec/controllers/cases/data_requests_controller_spec.rb:85]: do you want to check the cached_date_received?

File: spec/models/data_request_spec.rb [hunk 4]
```diff
@@ -235,4 +236,15 @@
...
+  describe '#request_date_present?' do
+    context 'when no request dates available' do
+      subject(:data_request) { build :data_request }
+      it { expect(subject.request_date_present?).to eq false }
+    end
+  end
+  describe '#request_dates_both_present?'
+  describe '#request_date_from_only?'
+  describe '#request_date_to_only?'
+  describe '#request_dates_absent?'
```
Comment (ymao2) [inline spec/models/data_request_spec.rb:249]: hmm, are you going to add test spec for those attributes?

File: app/views/cases/data_requests/_form.html.slim [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        form_hint_text: t('cases.data_requests.new.date_from_copy'),
+        # today_button: { class: '' } \
+      }
+
+      br
+
+      = f.gov_uk_date_field :date_to, { \
+        legend_text: 'Date to (optional)',
+        form_hint_text: t('cases.data_requests.new.date_to_copy'),
+        # today_button: { class: '' } \
```
Comment (willmcb) [inline app/views/cases/data_requests/_form.html.slim:39]: Do these commented out bits serve a purpose? 🤔

**Comment by willmcb** [comment]
General comment - should we be able to change the number of pages received after the request is closed?

File: app/helpers/cases_helper.rb [hunk 6]
```diff
@@ -276,14 +276,14 @@ def request_details_html(kase)
     content_tag(:strong, "#{kase.subject} ", class: 'strong') +
```
Comment (willmcb) [inline app/helpers/cases_helper.rb:276]: I think there is some refactoring of this method that could be done as we check `third_party?` a lot - have spoken to @njpreddy about this and we said we'd pair on it after the migration 👍

File: spec/models/data_request_spec.rb [hunk 7]
```diff
@@ -146,7 +147,7 @@
   describe '#request_type' do
     context 'valid values' do
       it 'does not error' do
-        expect(build(:data_request, request_type: 'offender')).to be_valid
+        expect(build(:data_request, request_type: 'all_prison_records')).to be_valid
```
Comment (willmcb) [inline spec/models/data_request_spec.rb:150]: There are quite a lot of `calls to `build(` in this file that I think we could refactor to `build_stubbed` which is a quicker in-memory alternative which doesn't call `create` and therefore avoids DB calls ([more info here](https://thoughtbot.com/blog/use-factory-girls-build-stubbed-for-a-faster-test#:~:text=build_stubbed%20is%20the%20younger%2C%20more,that's%20where%20the%20similarities%20end.)). 

I changed all of the builds to build stubbed in this file and the tests passed in 4.84 seconds, with build it took 6.66 seconds.

**Review by willmcb — APPROVED** [review]
Looks good 👍 - have made some non-blocking comments :)

File: db/migrate/20200819171428_drop_and_recreate_request_type_enum.rb (hunk 1 above)
Comment (Polsonby) [reply db/migrate/20200819171428_drop_and_recreate_request_type_enum.rb:4]: 🤦 sorry!

File: app/views/cases/data_requests/_form.html.slim [hunk 8]
```diff
@@ -0,0 +1,48 @@
+    .data-request__single.js-data-request__form
+      h3.page-heading--secondary.data-request__position
+
+      = f.text_field :location,
+          maxlength: 500,
+          class: 'js-data-request__location'
+
+      = f.radio_button_fieldset :request_type,
+          choices: DataRequest::request_types.keys do |fieldset|
```
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/views/cases/data_requests/_form.html.slim:9]: Yes, this is a bummer. The helper is amazing - just calling `f.radio_button_fieldset :request_type, choices: DataRequest::request_types.keys` does everything, it is a loop. But the problem is the shown/hidden textarea for "Other"; this explicit itemisation was the easiest way to output the hidden field beside the right element

File: spec/models/data_request_spec.rb (hunk 7 above)
Comment (Polsonby) [reply spec/models/data_request_spec.rb:150]: Interesting 🤔 I tried it and not all the objects get built correctly, and there wasn't any significant improvement in the run speed. But I'm really keen to explore this more 👍

**Review by ymao2 — APPROVED** [review]

**Pull request merged**